iroh-net/src/hp/magicsock/endpoint.rs

```diff
@@ -112,52 +116,122 @@ impl Endpoint {
-            derp_addr = self.derp_addr;
+    fn addr_for_send(&mut self, now: &Instant) -> (Option<SocketAddr>, Option<SocketAddr>, bool) {
+        match self.best_addr {
+            Some(ref best_addr) => {
+                if !self.is_best_addr_valid(*now) {
+                    // We had a best_addr but it expired so send both to it and DERP.
+                    debug!(
+                        "best addr is outdated {:?} {:?} - {:?}",
+                        now, best_addr, self.trust_best_addr_until
+                    );
+
+                    (Some(best_addr.addr), self.derp_addr, true)
+                } else {
+                    // Address is current and can be used
+                    (Some(best_addr.addr), None, false)
+                }
+            }
+            None => {
+                let (addr, should_ping) = self.get_candidate_udp_addr(now);
+
+                // provide backup derp addr if no known latency or no addr
+                let derp_addr = if should_ping || addr.is_none() {
+                    self.derp_addr
+                } else {
+                    None
+                };
+
+                (addr, derp_addr, should_ping)
+            }
+        }
+    }
+
+    /// Determines a potential best addr for this endpoint. And if the endpoint needs a ping.
+    fn get_candidate_udp_addr(&mut self, now: &Instant) -> (Option<SocketAddr>, bool) {
+        let mut lowest_latency = Duration::from_secs(60 * 60);
+        let mut last_pong = None;
+        for (ipp, state) in self.endpoint_state.iter() {
+            if let Some(pong) = state.recent_pong() {
+                // Lower latency, or when equal, prever IPv6.
+                if pong.latency < lowest_latency
+                    || (pong.latency == lowest_latency && ipp.is_ipv6())
+                {
+                    lowest_latency = pong.latency;
+                    last_pong.replace(pong);
+                }
+            }
+        }
+
+        // If we found a candidate, set to best addr
+        if let Some(pong) = last_pong {
+            self.best_addr = Some(AddrLatency {
+                addr: pong.from,
+                latency: Some(lowest_latency),
+            });
+            self.trust_best_addr_until
+                .replace(pong.pong_at + Duration::from_secs(60 * 60));
+
+            // No need to ping, we already have a latency.
+            return (Some(pong.from), false);
+        }
+
+        // Randomly select an address to use until we retrieve latency information.
+        let udp_addr = self
+            .endpoint_state
+            .keys()
+            .choose_stable(&mut rand::thread_rng())
+            .copied();
+        if let Some(addr) = udp_addr {
+            self.best_addr = Some(AddrLatency {
+                addr,
+                latency: None,
+            });
+
+            self.trust_best_addr_until = Some(*now + Duration::from_secs(15));
```

Comment:
Okay, so from looking at logs in the failing cli test, there are two main issues i'm seeing.

This issue is apparent on the get side, since as soon as we connect we try to send something. 

Issue One, that needs some more thinking:
The first time we attempt to do a full ping with a `callmemaybe` over derp is the first time we _ever_ connect to the derp server. This requires a full TLS handshake and takes some time to negotiate. But `callmemaybe` is sensitive to time. This is highly likely to never hole punch. 

Issue Two:
This is related to the actual code I'm commenting on 🤣 . Once we set the `best_addr` and `trust_best_addr`, we are potentially condemned to using a bad address for 15 seconds. Plus, since it's considered valid, we won't back it up with a derp address even though we have had no proof (yet) that it is a valid address.

One possible solution would be to keep `self.trust_best_addr_until` `None`, until a `full_ping` comes back successfully. But, this means that if another pong has come back on a different address during this time, we won't ever attempt to use it, because we still have a bad address as the `best_addr`.

I'm proposing, instead, that we never set an address as a `best_addr` unless it has a latency. 

So that would mean just removing these lines. (this fixes the bugs, locally for me, at least).

if we are worried about candidate address flip/flopping, we can keep track of a `candidate_addr`, that is the address we have randomly chosen as the default to use until we get any `pong`s back. We can loop through the addresses, and if none have any latency, we default to using the previous `candiate_addr`.

Reply:
> are potentially condemned to using a bad address for 15 seconds.

Well only for `5s`, as the pings get sent out and if we time out, it gets removed.

> Plus, since it's considered valid, 

this was the big problem, we drop derp because we think it is valid, forgot my own logic last night..